torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -20,13 +27,141 @@ class TritonKernelWrapperFunctional(HigherOrderOperator):
         super().__init__("triton_kernel_wrapper_functional")
 
-    def __call__(self, *, kernel, grid, kwargs):
-        kwargs = {
+
+triton_kernel_wrapper_functional = TritonKernelWrapperFunctional()
+
+
+@triton_kernel_wrapper_mutation.py_impl(DispatchKey.CompositeExplicitAutograd)
+def triton_kernel_wrapper_mutation_dense(*, kernel, grid, kwargs):
+    kernel[grid](**kwargs)
+
+
+@triton_kernel_wrapper_mutation.py_impl(FakeTensorMode)
+def triton_kernel_wrapper_mutation_fake_tensor_mode(mode, *, kernel, grid, kwargs):
+    with mode:
+        return None
+
+
+def trace_triton_kernel_wrapper(proxy_mode, func_overload, *, kernel, grid, kwargs):
+    with disable_proxy_modes_tracing():
+        out = func_overload(kernel=kernel, grid=grid, kwargs=kwargs)
+
+    fn = functools.partial(func_overload, kernel=kernel)
+    # FX graph needs __name__ and __module__ attributes
+    fn.__name__ = func_overload.__name__  # type:ignore[attr-defined]
+    if not hasattr(fn, "__module__"):
+        # Super hacky but on AMD __module__ is not set
+        fn.__module__ = "itertools"
+
+    node_args = {"grid": grid, "kwargs": kwargs}
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, node_args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function",
+        fn,
+        (),
+        proxy_args,
+        name="triton_kernel_wrapper_mutation",
+    )
+    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+
+@triton_kernel_wrapper_mutation.py_impl(ProxyTorchDispatchMode)
+def triton_kernel_wrapper_mutation_proxy_torch_dispatch_mode(
+    mode, *, kernel, grid, kwargs
+):
+    if mode.enable_tracing:
+        trace_triton_kernel_wrapper(
+            mode,
+            triton_kernel_wrapper_mutation,
+            kernel=kernel,
+            grid=grid,
+            kwargs=kwargs,
+        )
+    else:
+        triton_kernel_wrapper_mutation(kernel=kernel, grid=grid, kwargs=kwargs)
+
+    return None
+
+
+@triton_kernel_wrapper_mutation.py_functionalize_impl
+def triton_kernel_wrapper_mutation_functionalize(ctx, kernel, grid, kwargs):
+    unwrapped_kwargs = ctx.unwrap_tensors(kwargs)
+    with ctx.redispatch_to_next():
+        unwrapped_outputs = triton_kernel_wrapper_functional(
+            kernel=kernel, grid=grid, kwargs=unwrapped_kwargs
+        )
+
+    assert unwrapped_outputs.keys() == kwargs.keys()
+    for key, output_arg in unwrapped_outputs.items():
+        if not isinstance(output_arg, Tensor):
+            continue
+        input_arg = kwargs[key]
+        assert isinstance(input_arg, Tensor)
+
+        ctx.replace(input_arg, output_arg)
+        ctx.commit_update(input_arg)
+        ctx.sync(input_arg)
+    return None
+
+
+@triton_kernel_wrapper_functional.py_impl(DispatchKey.CompositeExplicitAutograd)
+def triton_kernel_wrapper_functional_dense(*, kernel, grid, kwargs):
+    kwargs = {
+        key: (clone_preserve_strides(val) if isinstance(val, Tensor) else val)
```

Comment:
One comment I made to Richard: maybe we should add a TODO here: eventually, for performance reasons we probably want to ensure that these `clone_preserve_strides` calls are **never** executed at runtime (inductor should always optimize them away).

We can't actually guarantee this today though, until we fix this issue: https://github.com/pytorch/pytorch/issues/109240